File: dotnet-impl/test/net/terria/registry/terraform/RegistryServiceTest.cs
using System;
using System.IO;
using Microsoft.Extensions.Configuration;
using Microsoft.Extensions.Logging;
using NUnit.Framework;
using net.terria.registry.terraform;

namespace net.terria.registry.terraform
{
    [TestFixture]
    public class RegistryServiceTest
    {
        private string _tempDir;

        [SetUp]
        public void Setup()
        {
            _tempDir = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString());
            Directory.CreateDirectory(_tempDir);
        }

        [TearDown]
        public void TearDown()
        {
            if (Directory.Exists(_tempDir))
            {
                Directory.Delete(_tempDir, true);
            }
        }

        [Test]
        public void TestInitWithEmptyStorage()
        {
            var config = new ConfigurationBuilder()
                .AddInMemoryCollection(new[] { new KeyValuePair<string, string>("registry.storage.path", _tempDir) })
                .Build();

            var logger = LoggerFactory.Create(builder => builder.AddConsole()).CreateLogger<RegistryService>();

            var service = new RegistryService(logger, config);

            Assert.That(service.GetProviders().Count, Is.EqualTo(0));
            Assert.That(service.GetModules().Count, Is.EqualTo(0));
        }

        [Test]
        public void TestInitWithProvider()
        {
            // Create provider structure under registry.terraform.io/providers
            var providersPath = Path.Combine(_tempDir, "registry.terraform.io", "providers");
            var namespacePath = Path.Combine(providersPath, "hashicorp");
            var providerPath = Path.Combine(namespacePath, "aws");
            var versionPath = Path.Combine(providerPath, "1.0.0");
            var platformPath = Path.Combine(versionPath, "linux", "amd64");

            Directory.CreateDirectory(platformPath);

            var config = new ConfigurationBuilder()
                .AddInMemoryCollection(new[] { new KeyValuePair<string, string>("registry.storage.path", _tempDir) })
                .Build();

            var logger = LoggerFactory.Create(builder => builder.AddConsole()).CreateLogger<RegistryService>();

            var service = new RegistryService(logger, config);

            Assert.That(service.GetProviders().Count, Is.EqualTo(1));
            Assert.That(service.GetProvider("registry.terraform.io", "hashicorp", "aws"), Is.Not.Null);
        }

        // Add more tests as needed
    }
}

File: java-impl/src/test/java/net/terria/registry/terraform/RegistryServiceTest.java
package net.terria.registry.terraform;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.UUID;

import static org.assertj.core.api.Assertions.assertThat;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.DynamicPropertyRegistry;
import org.springframework.test.context.DynamicPropertySource;

@SpringBootTest
public class RegistryServiceTest {

    private static final Path TEST_STORAGE_PATH = Path.of(System.getProperty("java.io.tmpdir"), "test-registry-" + UUID.randomUUID());

    @Autowired
    private RegistryService registryService;

    @DynamicPropertySource
    static void dynamicProperties(DynamicPropertyRegistry registry) {
        registry.add("registry.storage.path", () -> TEST_STORAGE_PATH.toString());
    }

    @Test
    public void testInitWithEmptyStorage() throws IOException {
        // Ensure path does not exist yet
        if (Files.exists(TEST_STORAGE_PATH)) {
            Files.walk(TEST_STORAGE_PATH)
                 .sorted((a, b) -> b.compareTo(a))
                 .forEach(path -> path.toFile().delete());
        }

        // Re-init service after cleanup
        registryService.init();

        assertThat(registryService.getProviders()).isEmpty();
        assertThat(registryService.getModules()).isEmpty();
    }

    @Test
    public void testInitWithProvider() throws IOException {
        // Create provider structure under registry.terraform.io/providers/
        Path registryPath = TEST_STORAGE_PATH.resolve("registry.terraform.io");
        Path providersPath = registryPath.resolve("providers");
        Path namespacePath = providersPath.resolve("hashicorp");
        Path providerPath = namespacePath.resolve("aws");
        Path versionPath = providerPath.resolve("1.0.0");
        Path platformPath = versionPath.resolve("linux").resolve("amd64");

        Files.createDirectories(platformPath);
        Files.createFile(providersPath.resolve("namespaces.json"));

        // Re-init service after test data creation
        registryService.init();

        assertThat(registryService.getProviders()).hasSize(1);
        assertThat(registryService.getProvider("registry.terraform.io", "hashicorp", "aws")).isPresent();
    }

    // Add more tests as needed
}